### PR TITLE
Preview provider registration in manager

### PR DESCRIPTION
**Pull request by nickvergessen**
As per #13648 (comment)

 @MorrisJobke @DeepDiver1975 @LukasReschke

### CODE REVIEW

File: lib/private/previewmanager.php [hunk 1]
```diff
@@ -21,7 +80,7 @@ class PreviewManager implements IPreview {
 	 * @param boolean $scaleUp Scale smaller images up to the thumbnail size or not. Might look ugly
 	 * @return \OCP\Image
```
Comment (nickvergessen) [inline lib/private/previewmanager.php]: `\OCP\Image` extends `\OC_Image` and `getPreview()` returns `\OC_Image` so this wrong.

However there is no `\OCP\IImage` (and it looks stupid), so I'd suggest to:
1. Add an interface `\OCP\IImage`
2. Make `\OC_Image` implement `\OCP\IImage` and ignore `\OCP\Image` (which then still implements the interface)
3. Change all type references from `\OC_Image` and `\OCP\Image` to `\OCP\IImage`

Opinions? @DeepDiver1975

File: lib/private/previewmanager.php [hunk 2]
```diff
@@ -32,17 +91,181 @@ function createPreview($file, $maxX = 100, $maxY = 75, $scaleUp = false) {
...
-		return \OC\Preview::isMimeSupported($mimeType);
+	public function isMimeSupported($mimeType = '*') {
+		if (!$this->config->getSystemValue('enable_previews', true)) {
+			return false;
+		}
+
+		$providerMimeTypes = array_keys($this->providers);
+		foreach ($providerMimeTypes as $supportedMimeType) {
+			if (preg_match($supportedMimeType, $mimeType)) {
+				return true;
```
Comment (nickvergessen) [inline lib/private/previewmanager.php]: performance improvement idea, cache this result

File: lib/private/previewmanager.php (hunk 1 above)
Comment (DeepDiver1975) [inline lib/private/previewmanager.php]: > Opinions? @DeepDiver1975

sounds reasonable .. go for it - thx

File: lib/private/previewmanager.php [hunk 3]
```diff
@@ -32,17 +91,181 @@ function createPreview($file, $maxX = 100, $maxY = 75, $scaleUp = false) {
...
+	 *  - OC\Preview\Font
+	 *
+	 * @return array
+	 */
+	protected function getEnabledDefaultProvider() {
+		if ($this->defaultProviders !== null) {
+			return $this->defaultProviders;
+		}
+
+		$this->defaultProviders = $this->config->getSystemValue('enabledPreviewProviders', [
```
Comment (LukasReschke) [inline lib/private/previewmanager.php]: I'd completely drop this then config option then and move these preview providers in apps. – We should only ship the ones we consider trusted then. (the ones enabled by default)

Others can be installed via AppStore etc. – No need to overcomplicate things here with an option switch :)

File: lib/private/previewmanager.php [hunk 4]
```diff
@@ -32,17 +91,181 @@ function createPreview($file, $maxX = 100, $maxY = 75, $scaleUp = false) {
...
+		$this->registerCoreProvider('OC\Preview\TXT', '/text\/plain/');
+		$this->registerCoreProvider('OC\Preview\MarkDown', '/text\/(x-)?markdown/');
+		$this->registerCoreProvider('OC\Preview\Image', '/image\/(?!tiff$)(?!svg.*).*/');
+		$this->registerCoreProvider('OC\Preview\MP3', '/audio\/mpeg/');
+
+		// SVG, Office and Bitmap require imagick
+		if (extension_loaded('imagick')) {
+			$checkImagick = new \Imagick();
+
+			$imagickProviders = [
```
Comment (LukasReschke) [inline lib/private/previewmanager.php]: Move all to third-party app (aka move to owncloud/apps)

File: lib/private/previewmanager.php [hunk 5]
```diff
@@ -32,17 +91,181 @@ function createPreview($file, $maxX = 100, $maxY = 75, $scaleUp = false) {
...
+				if (!in_array(trim($class, '\\'), $this->getEnabledDefaultProvider())) {
+					continue;
+				}
+
+				if (count($checkImagick->queryFormats($queryFormat)) === 1) {
+					$this->registerCoreProvider($class, $provider['mimetype']);
+				}
+			}
+
+			if (count($checkImagick->queryFormats('PDF')) === 1) {
```
Comment (LukasReschke) [inline lib/private/previewmanager.php]: Also 3rdparty

File: lib/private/previewmanager.php [hunk 6]
```diff
@@ -32,17 +91,181 @@ function createPreview($file, $maxX = 100, $maxY = 75, $scaleUp = false) {
...
+						$this->registerCoreProvider('\OC\Preview\OpenDocument', '/application\/vnd.oasis.opendocument.*/');
+						$this->registerCoreProvider('\OC\Preview\StarOffice', '/application\/vnd.sun.xml.*/');
+					}
+				}
+			}
+		}
+
+		// Video requires avconv or ffmpeg and is therefor
+		// currently not supported on Windows.
+		if (in_array('OC\Preview\Movie', $this->getEnabledDefaultProvider()) && !\OC_Util::runningOnWindows()) {
```
Comment (LukasReschke) [inline lib/private/previewmanager.php]: 3rdparty

**Comment by LukasReschke** [comment]
Good move! Thanks :rocket: 

Generally speaking we should drop all not officially supported providers and have the community handle them. We just cannot at the moment enable them without security impact etc. – And that whole "ffmpeg" magic and Co. drives me crazy :)

File: lib/private/previewmanager.php (hunk 3 above)
Comment (nickvergessen) [inline lib/private/previewmanager.php]: Next step

Comment (nickvergessen) [inline lib/private/previewmanager.php]: Should it be one app per provider, or one app and an admin section to enable/disable them individually?

Comment (DeepDiver1975) [inline lib/private/previewmanager.php]: > Should it be one app per provider

might be the easiest approach



> an admin section to enable/disable them individually

no gui from my pov - config.php should be enough - if necessary

File: apps/files_sharing/templates/public.php [hunk 7]
```diff
@@ -20,10 +20,10 @@
 OCP\Util::addscript('files', 'keyboardshortcuts');
 
 $thumbSize=1024;
-$previewSupported = OC\Preview::isMimeSupported($_['mimetype']) ? 'true' : 'false';
+$previewSupported = \OC::$server->getPreviewManager()->isMimeSupported($_['mimetype']) ? 'true' : 'false';
```
Comment (DeepDiver1975) [inline apps/files_sharing/templates/public.php]: remove from the template?

File: lib/private/previewmanager.php [hunk 8]
```diff
@@ -8,10 +8,87 @@
...
+	 * Constructor
+	 *
+	 * @param \OCP\IConfig $config
+	 */
+	public function __construct(\OCP\IConfig $config) {
+		$this->config = $config;
+
+		if ($this->config->getSystemValue('enable_previews', true)) {
+			// Register the default providers like txt, image, ...
+			$this->registerCoreProviders();
```
Comment (DeepDiver1975) [inline lib/private/previewmanager.php]: maybe do this a bit more lazy?

File: lib/private/previewmanager.php (hunk 3 above)
Comment (LukasReschke) [inline lib/private/previewmanager.php]: I'm also in favour of one or more providers per app. For example there might be an image_preview_provider app for all the basic formats and then a image_provider_raw app especially for RAW files.

Enabling the app should then just registee the provider. If the app is disabled the provider gets disabled.

File: lib/private/previewmanager.php (hunk 8 above)
Comment (nickvergessen) [inline lib/private/previewmanager.php]: done

File: apps/files_sharing/templates/public.php (hunk 7 above)
Comment (nickvergessen) [inline apps/files_sharing/templates/public.php]: done

File: lib/public/iimage.php [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+	 */
+	public function fixOrientation();
+
+	/**
+	 * Loads an image from a local file, a base64 encoded string or a resource created by an imagecreate* function.
+	 *
+	 * @param resource|string $imageRef The path to a local file, a base64 encoded string or a resource created by an imagecreate* function or a file resource (file handle    ).
+	 * @return resource|false An image resource or false on error
+	 */
+	public function load($imageRef);
```
Comment (DeepDiver1975) [inline lib/public/iimage.php]: from an API perspective I'd move all these load methods to a factory

File: lib/public/iimage.php [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+ * later.
+ * See the COPYING-README file.
+ */
+namespace OCP;
+
+
+/**
+ * Class for basic image manipulation
+ */
+interface IImage {
```
Comment (DeepDiver1975) [inline lib/public/iimage.php]: Are all these methods currently being used by potential callers of OCP\Preview? We shall be conservative and only expose those methods which are required.

File: lib/private/previewmanager.php (hunk 3 above)
Comment (oparoz) [inline lib/private/previewmanager.php]: @LukasReschke - How do you touch `lib/private/mimetypes.list.php` and `lib/repair/repairmimetypes.php` with an app?
How do you change that line with an app? https://github.com/oparoz/core/blob/raw-picture-preview/lib/private/preview/image.php#L16

File: lib/public/iimage.php (hunk 9 above)
Comment (nickvergessen) [inline lib/public/iimage.php]: The contacts app also calls them atm. I'd love to do that in a different PR then.
But I guess we can remove the load* from the interface and if someone needs them, they just continue to ignore the public interface for now, until the factory has been created.

File: lib/public/iimage.php (hunk 10 above)
Comment (nickvergessen) [inline lib/public/iimage.php]: Removed the `load*()` and some others. Only the "image manipulation" is now in the public api

File: lib/private/previewmanager.php (hunk 3 above)
Comment (nickvergessen) [inline lib/private/previewmanager.php]: * lib/private/mimetypes.list.php => you don't, send a request to core
* lib/repair/repairmimetypes.php => same here
* as for the mimetype on the image provider => we need to make it only accept what we want it to accept,
if it acccepts something an app accepts aswell, the one with the shorter mime type regex will be used, since it is asked first by the preview manager.

File: lib/private/preview.php [hunk 11]
```diff
@@ -95,11 +90,7 @@ public function __construct($user = '', $root = '/', $file = '', $maxX = 1, $max
 		$this->preview = null;
 
 		//check if there are preview backends
-		if (empty(self::$providers)) {
-			self::initProviders();
-		}
-
-		if (empty(self::$providers) && \OC::$server->getConfig()->getSystemValue('enable_previews', true)) {
+		if (!\OC::$server->getPreviewManager()->hasProviders() && \OC::$server->getConfig()->getSystemValue('enable_previews', true)) {
 			\OC_Log::write('core', 'No preview providers exist', \OC_Log::ERROR);
```
Comment (MorrisJobke) [inline lib/private/preview.php]: s/exist/exists/

File: lib/private/previewmanager.php (hunk 3 above)
Comment (oparoz) [inline lib/private/previewmanager.php]: Yep, came to the same conclusion regarding mimetypes.list.

Regarding the image provider, it need to be modelled after the bitmap group of classes so that these formats can be turned off as well in case another provider is chosen, so separate classes for each mimetype

File: lib/private/previewmanager.php (hunk 4 above)
Comment (oparoz) [inline lib/private/previewmanager.php]: Why keep any in core? Most rely on external libs and you're always exposing yourself to new threats when parsing user provided content.

**Comment by oparoz** [comment]
Also, for this to work, at least all the methods described in #12772 have to be included, otherwise apps will stop working because providers will be in an object they can't use and the one they can use doesn't include any providers.

**Comment by nickvergessen** [comment]
these methods are on \OCP\IImage, not on the preview manager.

**Comment by oparoz** [comment]
Nope, they're on Preview used both before generating the preview
and after to have access to it.


   FROM: Joas Schilling
SENT: Monday, 16 March 2015 16:56
TO: core@noreply.github.com
REPLY TO: owncloud/core
CC: Olivier Paroz
SUBJECT: Re: [core] Preview provider registration in manager (#14857)

> these methods are on \OCP\IImage, not on the preview manager.
>
> —
> Reply to this email directly or view it on GitHub[1].



Liens:
------
[1] https://github.com/owncloud/core/pull/14857#issuecomment-81757956

**Comment by nickvergessen** [comment]
Execatly, the Preview, not the PreviewManager. So this is a seperate thing and should be done in a separate PR ;)

**Comment by oparoz** [comment]
No, not on the returned preview which is an \OC\Image, but in \OC\Preview itself.

PreviewManager is the public interface to Preview, already providing methods such as isMimeSupported() and createPreview(), so that's the class all apps should use and that's the only class which will have access to providers once they all move outside of core, so that class needs to have all the methods of Preview in order to be a proper replacement.

**Comment by nickvergessen** [comment]
`PreviewManager` has a method `createPreview()`
This takes the `$file`, `$maxX`, `$maxY`, `$scaleUp` as an argument and then returns the `getPreview()` on the Preview object.
All those parameters are also parameters on the `getThumbnail()` method of the preview providers: https://github.com/owncloud/core/pull/14857/files#diff-d8470bc02ce010eabf8e08c208e64108R29
So I can't see what would be missing.

**Comment by oparoz** [comment]
What's missing are all the public methods. We can't possibly know of all their use. I can already give you three examples of files which are not compatible with the PreviewManager:
* Missing setKeepAspect() & showPreview() needed in [preview.png](https://github.com/owncloud/core/blob/preview-provider-registration-in-manager/core/ajax/preview.php#L37)
* Missing isCached() needed in [Picture](https://github.com/owncloud/gallery/blob/master/ajax/batch.php#L60)
* Missing getPreview() needed in [Gallery+](https://github.com/interfasys/galleryplus/blob/master/preview/preview.php#L165)

All these methods are listed in #12772

It might not be the responsibility of this PR to include them all, but you can't disable providers in core without offering access to them the same way they're accessed to today.

**Comment by nickvergessen** [comment]
Sorry, but you seem to mix up the preview providers with what is done on the preview/image later on.
So yes, that is still the task of another PR, if it needs to be done at all (and not lead to some fixes in the apps)

**Comment by oparoz** [comment]
Do you have a reference to an issue or PR regarding that preview/image programme of work?
All I'm saying is that you can't move providers in 8.1 and do the work on Preview in 8.2. The efforts need to be coordinated or things will break. We're 2 weeks away from the feature freeze.

And yes, it needs to be done as you can't just fix the apps.

**Comment by MorrisJobke** [comment]
@oparoz @nickvergessen Is this now the way we want to have it? does it work already? Can we start reviewing? What about @LukasReschke comments?

**Comment by oparoz** [comment]
The overall plan works for me
**Currently in this PR**
`Apps -> Preview -> PreviewManager -> Providers in lib/private`

What's left is to convert all providers into AppFramework apps and have something like this

`Apps -> Preview -> PreviewManager -> Provider apps`

but if I'm not mistaken, this can be done in another PR.

**Comment by MorrisJobke** [comment]
> but if I'm not mistaken, this can be done in another PR.

Yes

**Comment by MorrisJobke** [comment]
Just tested with a bunch of files and all previews still work :+1:

**Comment by nickvergessen** [comment]
yes, this PR is complete.
Next step is to move the providers out in suitable junks of apps

**Comment by DeepDiver1975** [comment]
:+1:
